apollo-runtime/src/main/java/com/apollographql/apollo/subscription/WebSocketSubscriptionTransport.kt

```diff
@@ -0,0 +1,166 @@
+    val webSocketListener = WebSocketListener(this)
+    check(this.webSocketListener.compareAndSet(null, webSocketListener)) {
+      "Already connected"
+    }
+    webSocket.set(webSocketConnectionFactory.newWebSocket(webSocketRequest, webSocketListener))
+  }
+
+  override fun disconnect(message: OperationClientMessage) {
+    webSocket.getAndSet(null)
+        ?.close(1001, message.serializeToJson())
+    release()
+  }
+
+  override fun send(message: OperationClientMessage) {
+    val socket = webSocket.get() ?: run {
+      callback.onFailure(IllegalStateException("Send attempted on closed connection"))
+      return
+    }
+    socket.send(message.serializeToJson())
+  }
+
+  internal fun onOpen() {
+    callback.onConnected()
+  }
+
+  internal fun onMessage(message: OperationServerMessage?) {
+    callback.onMessage(message)
+  }
+
+  internal fun onFailure(t: Throwable?) {
+    try {
+      callback.onFailure(t)
+    } finally {
+      release()
+    }
+  }
+
+  internal fun onClosed() {
+    try {
+      callback.onClosed()
+    } finally {
+      release()
+    }
+  }
+
+  private fun release() {
+    webSocketListener.getAndSet(null)?.release()
+    webSocket.set(null)
+  }
+
+  private fun OperationClientMessage.serializeToJson(): String =
+      toJsonString(
+        writePayloadAsJsonString = writePayloadAsJsonString,
+        extensions = extensions
+    )
+
+  internal class WebSocketListener(delegate: WebSocketSubscriptionTransport) : okhttp3.WebSocketListener() {
+    private val delegateRef = WeakReference(delegate)
+
+    override fun onOpen(webSocket: WebSocket, response: Response) {
+      delegateRef.get()?.onOpen()
+    }
+
+    override fun onMessage(webSocket: WebSocket, text: String) {
+      delegateRef.get()?.onMessage(OperationServerMessage.fromJsonString(text))
+    }
+
+    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
+      delegateRef.get()?.onFailure(t)
+    }
+
+    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
+      delegateRef.get()?.onClosed()
+    }
+
+    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
+      delegateRef.get()?.onClosed()
+    }
+
+    fun release() {
+      delegateRef.clear()
+    }
+  }
+
+  /**
+   * @param webSocketUrl The URL of the GraphQL WebSocket API
+   * @param webSocketConnectionFactory The [WebSocket.Factory] to use to create the websockets
+   * @param writePayloadAsJsonString If the payload should be serialized as a JSON String (default false). Set this to `true` for AWS
+   *                                 AppSync integrations.
+   * @param extensions Any additional extensions to when sending start commands. Must not be changed after this creating the factory.
+   */
+  class Factory @JvmOverloads constructor(
+      webSocketUrl: String,
+      private val webSocketConnectionFactory: WebSocket.Factory,
+      private val writePayloadAsJsonString: Boolean = false,
```

Comment:
I'm not very happy with the format of this change and I'm happy talk alternatives. 
I like being able to pass arbitrary extensions but not how this is triggered.

One alternative would be to introduce some sort of "format" or "dialect" enum which there are 2 off. I tried to find some official specification for the websocket format but there doesn't seem to be one?